app/code/Magento/Catalog/Block/Product/ListProduct.php

```diff
@@ -516,7 +517,9 @@ private function configureToolbar(Toolbar $toolbar, Collection $collection)
             $toolbar->setModes($modes);
         }
         // set collection to toolbar and apply sort
-        $toolbar->setCollection($collection);
+        if (!$toolbar->getCollection()) {
```

Comment:
I think it's not really logical to solve issues with initializing toolbar object in the client code. In case if someone use it in custom module - issue will appear again. I think it should be moved inside Toolbar class as internal logic.

I'm not totally sure what is the best solution would be there, but not setting collection not looks logically here.

Could you describe with more details how issue actually appears?